File: mapper/1527_mapper_server.py
#!/usr/bin/env python3
import os
"""
1527 Mapper - HTTP API Gateway
Relays MQTT messages between browser and ESP32 gateway
"""

import asyncio
import json
import threading
from flask import Flask, request, jsonify, Response, send_file
from gevent.pywsgi import WSGIServer
import paho.mqtt.client as mqtt

app = Flask(__name__)

# MQTT Settings
MQTT_BROKER = os.environ.get("MQTT_BROKER", "your-mqtt-broker.example.com")
MQTT_PORT = int(os.environ.get("MQTT_PORT", "1883"))
MQTT_USER = os.environ.get("MQTT_USER", "your-mqtt-username")
MQTT_PASS = os.environ.get("MQTT_PASS", "your-mqtt-password")
MQTT_TOPIC_PUB = "home/gateway/1527map"
MQTT_TOPIC_SUB = "home/gateway/1527map_resp"

# Global state
last_response = {"data": None, "timestamp": None}
response_event = threading.Event()


def on_connect(client, userdata, flags, rc, properties=None):
    if rc == 0:
        print(f"[MQTT] Connected to broker")
        client.subscribe(MQTT_TOPIC_SUB, qos=1)
    else:
        print(f"[MQTT] Connection failed with code {rc}")


def on_message(client, userdata, msg):
    try:
        payload = json.loads(msg.payload.decode())
        print(f"[MQTT] Received on {msg.topic}: {payload}")
        last_response["data"] = payload
        last_response["timestamp"] = asyncio.get_event_loop().time() if asyncio.get_event_loop().is_running() else 0
        response_event.set()
        response_event.clear()
    except Exception as e:
        print(f"[MQTT] Error parsing message: {e}")


def on_disconnect(client, userdata, rc, properties=None):
    print(f"[MQTT] Disconnected, reconnecting...")


# MQTT client setup (singleton)
mqtt_client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2)
mqtt_client.username_password_set(MQTT_USER, MQTT_PASS)
mqtt_client.on_connect = on_connect
mqtt_client.on_message = on_message
mqtt_client.on_disconnect = on_disconnect
mqtt_client.connect(MQTT_BROKER, MQTT_PORT, keepalive=60)
mqtt_client.loop_start()


# --- API Routes ---

@app.route("/")
def index():
    return send_file("index.html")


@app.route("/api/status")
def status():
    """Check MQTT connection status"""
    connected = mqtt_client.is_connected()
    return jsonify({
        "mqtt_connected": connected,
        "last_response": last_response["data"]
    })


@app.route("/api/send", methods=["POST"])
def send():
    """Publish a message to MQTT topic"""
    if not mqtt_client.is_connected():
        return jsonify({"error": "MQTT not connected"}), 500

    data = request.get_json()
    if not data:
        return jsonify({"error": "No data provided"}), 400

    payload = json.dumps(data)
    result = mqtt_client.publish(MQTT_TOPIC_PUB, payload, qos=1)
    
    if result.rc == mqtt.MQTT_ERR_SUCCESS:
        print(f"[API] Published to {MQTT_TOPIC_PUB}: {data}")
        return jsonify({"success": True, "message_id": result.mid})
    else:
        return jsonify({"error": f"MQTT error code {result.rc}"}), 500


@app.route("/api/subscribe")
def subscribe():
    """Server-Sent Events endpoint for receiving MQTT responses"""
    def generate():
        # Send initial connection确认
        yield f"data: {json.dumps({'type': 'connected'})}\n\n"
        
        while True:
            response_event.wait(timeout=30)
            if last_response["data"]:
                yield f"data: {json.dumps({'type': 'response', 'data': last_response['data']})}\n\n"
                last_response["data"] = None

    return Response(generate(), mimetype="text/event-stream",
                    headers={"Cache-Control": "no-cache",
                             "Connection": "keep-alive",
                             "X-Accel-Buffering": "no"})


@app.route("/api/load")
def load():
    """Load current map from ESP32"""
    if not mqtt_client.is_connected():
        return jsonify({"error": "MQTT not connected"}), 500
    
    result = mqtt_client.publish(MQTT_TOPIC_PUB, json.dumps({"cmd": "get_map"}), qos=1)
    return jsonify({"success": result.rc == mqtt.MQTT_ERR_SUCCESS})


if __name__ == "__main__":
    import os
    work_dir = os.path.dirname(os.path.abspath(__file__))
    
    print(f"[1527 Mapper] Starting HTTP server on :18597")
    print(f"[1527 Mapper] MQTT broker: {MQTT_BROKER}:{MQTT_PORT}")
    print(f"[1527 Mapper] Publish topic: {MQTT_TOPIC_PUB}")
    print(f"[1527 Mapper] Subscribe topic: {MQTT_TOPIC_SUB}")
    
    http_server = WSGIServer(("0.0.0.0", 18597), app)
    http_server.serve_forever()

File: mapper/1527_mapper_server.js
const express = require('express');
const mqtt = require('mqtt');
const fs = require('fs');
const path = require('path');

const app = express();
const PORT = 18597;

// MQTT Settings
const MQTT_BROKER = 'mqtt://127.0.0.1:1883';
const MQTT_USER = process.env.MQTT_USER || 'YOUR_MQTT_USERNAME';
const MQTT_PASS = process.env.MQTT_PASS || 'YOUR_MQTT_PASSWORD';
const PUB_TOPIC = 'home/gateway/1527map';
const SUB_TOPIC = 'home/gateway/1527map_resp';

// State
let lastResponse = null;
const clients = []; // SSE clients

// Middleware
app.use(express.json());
app.use(express.static(__dirname));

// Connect to MQTT
const mqttClient = mqtt.connect(MQTT_BROKER, {
  username: MQTT_USER,
  password: MQTT_PASS,
  keepalive: 60,
  reconnectPeriod: 3000,
});

mqttClient.on('connect', () => {
  console.log('[MQTT] Connected to broker');
  mqttClient.subscribe(SUB_TOPIC, { qos: 1 });
});

mqttClient.on('message', (topic, message) => {
  try {
    const payload = JSON.parse(message.toString());
    console.log('[MQTT] Received:', payload);
    lastResponse = payload;
    // Notify all SSE clients
    clients.forEach(client => {
      client.write(`data: ${JSON.stringify({ type: 'response', data: payload })}\n\n`);
    });
  } catch (e) {
    console.error('[MQTT] Parse error:', e.message);
  }
});

mqttClient.on('error', (err) => {
  console.error('[MQTT] Error:', err.message);
});

mqttClient.on('offline', () => {
  console.log('[MQTT] Offline');
});

// SSE endpoint
app.get('/api/subscribe', (req, res) => {
  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.setHeader('X-Accel-Buffering', 'no');
  res.flushHeaders();

  // Send initial connected message
  res.write(`data: ${JSON.stringify({ type: 'connected' })}\n\n`);
  console.log('[SSE] Client connected');

  clients.push(res);

  // Keep-alive ping
  const ping = setInterval(() => {
    res.write(`: ping\n\n`);
  }, 30000);

  req.on('close', () => {
    clearInterval(ping);
    const idx = clients.indexOf(res);
    if (idx > -1) clients.splice(idx, 1);
    console.log('[SSE] Client disconnected');
  });
});

// API: Check status
app.get('/api/status', (req, res) => {
  res.json({
    mqtt_connected: mqttClient.connected,
    last_response: lastResponse
  });
});

// API: Publish message
app.post('/api/send', (req, res) => {
  if (!mqttClient.connected) {
    return res.status(500).json({ error: 'MQTT not connected' });
  }
  const { cmd, data } = req.body;
  const payload = JSON.stringify({ cmd, data });
  mqttClient.publish(PUB_TOPIC, payload, { qos: 1 }, (err) => {
    if (err) {
      console.error('[API] Publish error:', err);
      return res.status(500).json({ error: err.message });
    }
    console.log('[API] Published:', payload);
    res.json({ success: true });
  });
});

// API: Request map from ESP32
app.get('/api/load', (req, res) => {
  if (!mqttClient.connected) {
    return res.status(500).json({ error: 'MQTT not connected' });
  }
  mqttClient.publish(PUB_TOPIC, JSON.stringify({ cmd: 'get_map' }), { qos: 1 });
  res.json({ success: true });
});

// Start server
app.listen(PORT, '0.0.0.0', () => {
  console.log(`[1527 Mapper] Server running on http://0.0.0.0:${PORT}`);
  console.log(`[1527 Mapper] MQTT: ${MQTT_BROKER}`);
  console.log(`[1527 Mapper] Topics: ${PUB_TOPIC} / ${SUB_TOPIC}`);
});
